Cinergy S2 Box ID 0ccd:0105 with Montage-3103b

Hi, my device is TerraTec Electronic GmbH Cinergy S2 Box ID 0ccd:0105 with Montage-3103b

informations on kernel without drivers (media_build)

lsusb:

Bus 001 Device 002: ID 0ccd:0105 TerraTec Electronic GmbH Cinergy S2 Box

dmesg :

[ 1249.760585] usb 1-1: new high-speed USB device number 4 using ehci-pci
[ 1249.890023] usb 1-1: New USB device found, idVendor=0ccd, idProduct=0105, bcdDevice= 0.00
[ 1249.890026] usb 1-1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[ 1249.890029] usb 1-1: Product: TERRATEC USB Tuner
[ 1249.890031] usb 1-1: Manufacturer: Tevii
[ 1249.890033] usb 1-1: SerialNumber: 000000031

informations on kernel with drivers (media_tree) from linuxtv brad repo

ls -l /dev/dvb/adapter0 :

razem 0
crw-rw----+ 1 root video 212, 4 11-08 15:59 demux0
crw-rw----+ 1 root video 212, 5 11-08 15:59 dvr0
crw-rw----+ 1 root video 212, 7 11-08 15:59 net0

frontend is missing :

    "[ 33.279671] m88ds3103 0-0068: Unknown device. Chip_id=52"

[ 27.252906] usb 1-1: USB disconnect, device number 2
[ 27.262960] dvb-usb: Terratec Cinergy S2 USB BOX successfully deinitialized and disconnected.
[ 32.830356] usb 1-1: new high-speed USB device number 4 using ehci-pci
[ 32.959923] usb 1-1: New USB device found, idVendor=0ccd, idProduct=0105, bcdDevice= 0.00
[ 32.959926] usb 1-1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[ 32.959929] usb 1-1: Product: TERRATEC USB Tuner
[ 32.959931] usb 1-1: Manufacturer: Tevii
[ 32.959933] usb 1-1: SerialNumber: 000000031
[ 32.960354] dw2102: su3000_identify_state
[ 32.960355] dvb-usb: found a 'Terratec Cinergy S2 USB BOX' in warm state.
[ 32.960365] dw2102: su3000_power_ctrl: 1, initialized 0
[ 32.962060] dvb-usb: will pass the complete MPEG2 transport stream to the software demuxer.
[ 32.962188] dvbdev: DVB: registering new adapter (Terratec Cinergy S2 USB BOX)
[ 32.962194] usb 1-1: media controller created
[ 32.966918] dvb-usb: MAC address: bc:ea:2b:46:13:a5
[ 32.967297] dvbdev: dvb_create_media_entity: media entity 'dvb-demux' registered.

    [ 33.279671] m88ds3103 0-0068: Unknown device. Chip_id=52
    [ 33.279778] dvb-usb: no frontend was attached by 'Terratec Cinergy S2 USB BOX'
    [ 33.280302] Registered IR keymap rc-tt-1500
    [ 33.280365] rc rc0: Terratec Cinergy S2 USB BOX as /devices/pci0000:00/0000:00:04.1/usb1/1-1/rc/rc0
    [ 33.280445] input: Terratec Cinergy S2 USB BOX as /devices/pci0000:00/0000:00:04.1/usb1/1-1/rc/rc0/input6
    [ 33.281696] rc rc0: lirc_dev: driver dw2102 registered at minor = 0, scancode receiver, no transmitter
    [ 33.281698] dvb-usb: schedule remote query interval to 250 msecs.
    [ 33.281701] dw2102: su3000_power_ctrl: 0, initialized 1
    [ 33.281702] dvb-usb: Terratec Cinergy S2 USB BOX successfully initialized and connected.

lsmod :

       Module                  Size  Used by
       dvb_usb_dvbsky         20480  0
       dvb_usb_v2             28672  1 dvb_usb_dvbsky
       rc_tt_1500             16384  0

            m88ds3103 28672 1 dvb_usb_dvbsky
            i2c_mux 16384 1 m88ds3103
            dvb_usb_dw2102 69632 0 <<< 0
            dvb_usb 28672 1 dvb_usb_dw2102
            dvb_core 126976 dvb_usb_v2,m88ds3103,dvb_usb,dvb_usb_dw2102,dvb_usb_dvbsky
            rc_core 40960 6 dvb_usb_v2,dvb_usb,dvb_usb_dw2102,dvb_usb_dvbsky,rc_tt_1500
            mc 40960 3 dvb_usb_v2,dvb_usb,dvb_core

module dvb_usb_dw2102 is not used with module m88ds3103 !
module m88ds3103 is used with module dvb_usb_dvbsky !

is there any option to pass m88ds3103 module load for 3103b version ?
is this lsmod is correct because dvb_usb_dw2102 is not used with module m88ds3103 ?
why this device is recognized at once by dw2102 and dvbsky/ blacklist first or second ?

is there any chance to add support for this device to your brantch Montage_3103b ?
or is already included but something is not configured ?

https://git.linuxtv.org/brad/media_tree.git/log/?h=Montage-3103b.v2

with upstream media_tree have warning:

  CC [M]  drivers/media/dvb-frontends/m88ds3103.o
  CC      lib/timerqueue.o
  CC      lib/vsprintf.o
drivers/media/dvb-frontends/m88ds3103.c: In function ‘m88ds3103_set_frontend’:
drivers/media/dvb-frontends/m88ds3103.c:906:6: warning: this statement may fall through [-Wimplicit-fallthrough=]
   if (dev->chiptype == M88DS3103_CHIPTYPE_3103B) {
      ^
drivers/media/dvb-frontends/m88ds3103.c:911:2: note: here
  default:
  ^~~~~~~
  CC [M]  drivers/media/dvb-frontends/mn88472.o
  CC [M]  drivers/media/dvb-frontends/mn88473.o

--- a/drivers/media/dvb-frontends/m88ds3103.c
+++ b/drivers/media/dvb-frontends/m88ds3103.c
@@ -1047,7 +1047,7 @@ static int m88ds3103_init(struct dvb_frontend *fe)
 	if (utmp)
 		goto warm;
 
-	/* global reset, global diseqc reset, golbal fec reset */
+	/* global reset, global diseqc reset, global fec reset */
 	ret = regmap_write(dev->regmap, 0x07, 0xe0);
 	if (ret)
 		goto err;

--- a/drivers/media/usb/dvb-usb/dw2102.c
+++ b/drivers/media/usb/dvb-usb/dw2102.c
@@ -1527,6 +1527,28 @@ static int m88rs2000_frontend_attach(struct dvb_usb_adapter *adap)
 	return -EIO;
 }
 
+static int tt_s2_4600_frontend_attach_probe_demod(struct dvb_usb_device *d, const int probe_addr)
+{
+	struct dw2102_state *state = d->priv;
+
+	state->data[0] = 0x9;
+	state->data[1] = 0x1;
+	state->data[2] = 0x1;
+	state->data[3] = probe_addr;
+	state->data[4] = 0x0;
+
+	if (dvb_usb_generic_rw(d, state->data, 5, state->data, 2, 0) < 0)
+	{
+		err("i2c probe for address 0x%x failed.", probe_addr);
+		return 0;
+	}
+	if(state->data[0] != 8)
+		/* failure (7) or error, no device at this address */
+		return 0;
+	/* probing successful */
+	return 1;
+}
+
 static int tt_s2_4600_frontend_attach(struct dvb_usb_adapter *adap)
 {
 	struct dvb_usb_device *d = adap->dev;
@@ -1536,6 +1558,7 @@ static int tt_s2_4600_frontend_attach(struct dvb_usb_adapter *adap)
 	struct i2c_board_info board_info;
 	struct m88ds3103_platform_data m88ds3103_pdata = {};
 	struct ts2020_config ts2020_config = {};
+	int demod_addr;
 
 	mutex_lock(&d->data_mutex);
 
@@ -1583,7 +1606,22 @@ static int tt_s2_4600_frontend_attach(struct dvb_usb_adapter *adap)
 	if(state->data[0] != 1)
 		err("command 0x51 reply unexpected.");
 
+	/* probe for demodulator i2c address */
+	demod_addr = -1;
+	if(tt_s2_4600_frontend_attach_probe_demod(d, 0x68))
+		demod_addr = 0x68;
+	else if(tt_s2_4600_frontend_attach_probe_demod(d, 0x69))
+		demod_addr = 0x69;
+	else if(tt_s2_4600_frontend_attach_probe_demod(d, 0x6a))
+		demod_addr = 0x6a;
+
 	mutex_unlock(&d->data_mutex);
+
+	if(demod_addr < 0)
+	{
+		err("probing for demodulator failed.  Is the external power switched on?");
+		return -ENODEV;
+	}
 
 	/* attach demod */
 	m88ds3103_pdata.clk = 27000000;
@@ -1599,8 +1637,11 @@ static int tt_s2_4600_frontend_attach(struct dvb_usb_adapter *adap)
 	m88ds3103_pdata.lnb_hv_pol = 1;
 	m88ds3103_pdata.lnb_en_pol = 0;
 	memset(&board_info, 0, sizeof(board_info));
-	strscpy(board_info.type, "m88ds3103", I2C_NAME_SIZE);
-	board_info.addr = 0x68;
+	if(demod_addr == 0x6a)
+		strscpy(board_info.type, "m88ds3103b", I2C_NAME_SIZE);
+	else
+		strscpy(board_info.type, "m88ds3103", I2C_NAME_SIZE);
+	board_info.addr = demod_addr;
 	board_info.platform_data = &m88ds3103_pdata;
 	request_module("m88ds3103");
 	client = i2c_new_device(&d->i2c_adap, &board_info);

--- a/drivers/media/usb/dvb-usb/dw2102.c
+++ b/drivers/media/usb/dvb-usb/dw2102.c
@@ -770,6 +770,9 @@ static int su3000_i2c_transfer(struct i2c_adapter *adap, struct i2c_msg msg[],
 					state->data, msg[1].len + 1, 0) < 0)
 			err("i2c transfer failed.");
 
+		if (state->data[0] != 8)
+		   warn("i2c read request failed: i2c status %d", state->data[0]);
+
 		memcpy(msg[1].buf, &state->data[1], msg[1].len);
 		break;
 	default:

--- a/drivers/media/usb/dvb-usb/dw2102.c
+++ b/drivers/media/usb/dvb-usb/dw2102.c
@@ -1544,14 +1544,18 @@ static int tt_s2_4600_frontend_attach(struct dvb_usb_adapter *adap)
 	state->data[2] = 0x0;
 
 	if (dvb_usb_generic_rw(d, state->data, 3, state->data, 1, 0) < 0)
-		err("command 0x0e transfer failed.");
+		err("command 0x0e 80 00 transfer failed.");
+	if(state->data[0] != 1)
+		err("command 0x0e 80 00 reply unexpected.");
 
 	state->data[0] = 0xe;
 	state->data[1] = 0x02;
 	state->data[2] = 1;
 
 	if (dvb_usb_generic_rw(d, state->data, 3, state->data, 1, 0) < 0)
-		err("command 0x0e transfer failed.");
+		err("command 0x0e 02 01 transfer failed.");
+	if(state->data[0] != 1)
+		err("command 0x0e 02 01 reply unexpected.");
 	msleep(300);
 
 	state->data[0] = 0xe;
@@ -1559,19 +1563,25 @@ static int tt_s2_4600_frontend_attach(struct dvb_usb_adapter *adap)
 	state->data[2] = 0;
 
 	if (dvb_usb_generic_rw(d, state->data, 3, state->data, 1, 0) < 0)
-		err("command 0x0e transfer failed.");
+		err("command 0x0e 83 00 transfer failed.");
+	if(state->data[0] != 1)
+		err("command 0x0e 83 00 reply unexpected.");
 
 	state->data[0] = 0xe;
 	state->data[1] = 0x83;
 	state->data[2] = 1;
 
 	if (dvb_usb_generic_rw(d, state->data, 3, state->data, 1, 0) < 0)
-		err("command 0x0e transfer failed.");
+		err("command 0x0e 83 01 transfer failed.");
+	if(state->data[0] != 1)
+		err("command 0x0e 83 01 reply unexpected.");
 
 	state->data[0] = 0x51;
 
 	if (dvb_usb_generic_rw(d, state->data, 1, state->data, 1, 0) < 0)
 		err("command 0x51 transfer failed.");
+	if(state->data[0] != 1)
+		err("command 0x51 reply unexpected.");
 
 	mutex_unlock(&d->data_mutex);
 

--- a/drivers/media/usb/dvb-usb/dw2102.c
+++ b/drivers/media/usb/dvb-usb/dw2102.c
@@ -1643,6 +1643,9 @@ static int tt_s2_4600_frontend_attach(struct dvb_usb_adapter *adap)
 		strscpy(board_info.type, "m88ds3103", I2C_NAME_SIZE);
 	board_info.addr = demod_addr;
 	board_info.platform_data = &m88ds3103_pdata;
+
+	info("%s: attaching demodulator of type %s at i2c address 0x%x", __func__, board_info.type, board_info.addr);
+
 	request_module("m88ds3103");
 	client = i2c_new_device(&d->i2c_adap, &board_info);
 	if (client == NULL || client->dev.driver == NULL)

